processor/src/test/java/org/mapstruct/ap/test/accessingmappers/InventoryMapper.java

```diff
@@ -0,0 +1,49 @@
+/**
+ *  Copyright 2012-2016 Gunnar Morling (http://www.gunnarmorling.de/)
+ *  and/or other contributors as indicated by the @authors tag. See the
+ *  copyright.txt file in the distribution for a full listing of all
+ *  contributors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.mapstruct.ap.test.accessingmappers;
+
+import org.mapstruct.Mapper;
+import org.mapstruct.factory.Mappers;
+
+
+/**
+ *
+ * @author Sjaak Derksen
+ */
+@Mapper(uses = { RadioMapper.class, TvMapper.class } )
+public abstract class InventoryMapper {
+
+    public static final InventoryMapper INSTANCE = Mappers.getMapper( InventoryMapper.class );
+
+    private final InventorySequence inventorySequence = new InventorySequence();
+
+    public  InventoryEntity mapInventory(InventoryDto source) {
+        getRadioMapper().init( inventorySequence );
```

Comment:
What about splitting the problem in two. What I mean is that the construction of abstract classes with overloaded constructors can be done by generating a factory from an interface with methods mimicking the constructors you want to use. E.g via our own @Factorize annotation. The factory itself can be constructed in the same manner as mappers.

Second part: a similar mechanism for zero arg constructors in mappers. The mapper could look for zero-arg method marked with our own @produces annotation. No implementation will be generated. The user has to implement those and call the proper method on the generated factory. 

Naming could be better probably.. 😄

Verstuurd vanaf mijn iPhone

> Op 25 feb. 2016 om 23:57 heeft Gunnar Morling <notifications@github.com> het volgende geschreven:
> 
> In processor/src/test/java/org/mapstruct/ap/test/accessingmappers/InventoryMapper.java:
> 
> > +import org.mapstruct.factory.Mappers;
> > +
> > +
> > +/**
> > + *
> > + * @author Sjaak Derksen
> > + */
> > +@Mapper(uses = { RadioMapper.class, TvMapper.class } )
> > +public abstract class InventoryMapper {
> > +
> > +    public static final InventoryMapper INSTANCE = Mappers.getMapper( InventoryMapper.class );
> > +
> > +    private final InventorySequence inventorySequence = new InventorySequence();
> > +
> > +    public  InventoryEntity mapInventory(InventoryDto source) {
> > +        getRadioMapper().init( inventorySequence );
> Yes, I understand where you coming from. I think it's a valid and common issue, but not one easily solved. Essentially, you need to share "state" during the runtime of one getMapper() call.
> 
> Just thinking out loud here: How about we provide an overloaded version of getMapper(), that takes a callback provided by the user which is in charge of instantiating required dependencies. That you could ensure identity of dependencies as needed:
> 
> Sequence sequence = ...;
> InventoryMapper inventoryMapper = Mappers.getMapper(
>     InventoryMapper.class,
>     ( dependencyClazz, defaultFactory ) -> {
>         if ( dependencyClazz = Sequence.class ) {
>             return sequence; // return same sequence instance during this call 
>         }
>         else {
>             return defaultFactory.getMapper( dependencyClazz );
>         }
>     }
> );
> The implementation would have to keep that callback as thread-local so it can be accessed by all getMapper() calls in the course of one invocation, but that'd be encapsulated within MapStruct's code of the factory, so the user wouldn't have to deal with that.
> 
> —
> Reply to this email directly or view it on GitHub.
>